nixos/modules/services/networking/unbound.nix

```diff
@@ -106,62 +88,150 @@ in
+      };
+      remote-control = {
+        control-enable = mkDefault false;
+        control-interface = mkDefault ([ "127.0.0.1" ] ++ (optional config.networking.enableIPv6 "::1"));
+        server-key-file = mkDefault "${cfg.stateDir}/unbound_server.key";
+        server-cert-file = mkDefault "${cfg.stateDir}/unbound_server.pem";
+        control-key-file = mkDefault "${cfg.stateDir}/unbound_control.key";
+        control-cert-file = mkDefault "${cfg.stateDir}/unbound_control.pem";
+      } // optionalAttrs (cfg.localControlSocketPath != null) {
+        control-enable = true;
+        control-interface = cfg.localControlSocketPath;
+      };
+    };
+
     environment.systemPackages = [ cfg.package ];
 
-    users.users.unbound = {
-      description = "unbound daemon user";
-      isSystemUser = true;
-      group = lib.mkIf (cfg.localControlSocketPath != null) (lib.mkDefault "unbound");
+    users.users = mkIf (cfg.user == "unbound") {
+      unbound = {
+        description = "unbound daemon user";
+        isSystemUser = true;
+        group = cfg.group;
+      };
     };
 
-    # We need a group so that we can give users access to the configured
-    # control socket. Unbound allows access to the socket only to the unbound
-    # user and the primary group.
-    users.groups = lib.mkIf (cfg.localControlSocketPath != null) {
+    users.groups = mkIf (cfg.group == "unbound") {
       unbound = {};
     };
 
-    networking.resolvconf.useLocalResolver = mkDefault true;
+    networking = mkIf cfg.resolveLocalQueries {
+      nameservers = [ "127.0.0.1" ] ++ (optional config.networking.enableIPv6 "::1");
 
+      resolvconf = {
+        useLocalResolver = mkDefault true;
+      };
+
+      networkmanager.dns = "unbound";
+    };
 
     environment.etc."unbound/unbound.conf".source = confFile;
 
+    systemd.tmpfiles.rules = [
+      "d '${cfg.stateDir}' 0755 ${cfg.user} nogroup - -"
+      "d '${cfg.stateDir}/dev' 0755 ${cfg.user} nogroup - -"
+      "Z '${cfg.stateDir}' - ${cfg.user} nogroup - -"
+    ];
+
     systemd.services.unbound = {
       description = "Unbound recursive Domain Name Server";
       after = [ "network.target" ];
       before = [ "nss-lookup.target" ];
       wantedBy = [ "multi-user.target" "nss-lookup.target" ];
 
-      preStart = lib.mkIf cfg.enableRootTrustAnchor ''
-        ${cfg.package}/bin/unbound-anchor -a ${rootTrustAnchorFile} || echo "Root anchor updated!"
+      path = mkIf cfg.settings.remote-control.control-enable [ pkgs.openssl ];
+
+      preStart = ''
+        ${optionalString cfg.enableRootTrustAnchor ''
+          ${cfg.package}/bin/unbound-anchor -a ${rootTrustAnchorFile} || echo "Root anchor updated!"
+        ''}
+        ${optionalString cfg.settings.remote-control.control-enable ''
+          ${cfg.package}/bin/unbound-control-setup -d ${cfg.stateDir}
+        ''}
       '';
 
       restartTriggers = [
-        confFile
+        "/etc/unbound/unbound.conf"
```

Comment:
Does this still properly trigger? I thought that field had to change in order to trigger a restart. If we only have that static path here it will likely never trigger a restart unless the binary changes?

Reply:
Well, the file is linked to `/etc/unbound/unbound.conf`, so it will work either way. However, we can consider making it `/etc/unbound` for extra manual configuration.

Reply:
@andir is correct - static strings do nothing here. `confFile` provides value because it changes, but you should never use a static string in this field.

Reply:
Will put it back then

Reply:
@rissson specifically what happens, when running `nixos-rebuild`, is that a hash of the `systemd` unit file is taken from your current generation and compared against the hash from the `systemd` unit in the new generation. If the hashes are different, which they would be if the path of your configuration file (`confFile`) has changed, then `nixos` knows to restart/reload the unit in question. Hence a static string never accomplishes anything.

Reply:
Okay, thanks for that explanation!

Reply:
Should we maybe use `reloadIfChanged` then?

Reply:
I'll ping @flokli on this one... maybe this is another reminder that the two of us need to meet up and talk about this again.

Reply:
Let me link to https://github.com/NixOS/nixpkgs/issues/49528#issuecomment-633097264, which should describe how we could be doing this.